controllers/openstackmachine_controller.go

```diff
@@ -230,15 +230,20 @@ func (r *OpenStackMachineReconciler) reconcileDelete(ctx context.Context, logger
 		return ctrl.Result{}, nil
 	}
 
-	if instance.Name != "" {
-		if err = computeService.DeleteInstance(openStackMachine, instance); err != nil {
-			handleUpdateMachineError(logger, openStackMachine, errors.Errorf("error deleting Openstack instance: %v", err))
-			return ctrl.Result{}, nil
-		}
+	if err = computeService.DeleteInstance(openStackMachine, instance.InstanceIdentifier()); err != nil {
+		handleUpdateMachineError(logger, openStackMachine, errors.Errorf("error deleting OpenStack instance %s with ID %s: %v", instance.Name(), instance.ID(), err))
+		return ctrl.Result{}, nil
+	}
+
+	ns, err := instance.NetworkStatus()
+	if err != nil {
+		handleUpdateMachineError(logger, openStackMachine, errors.Errorf("error getting network status for OpenStack instance %s with ID %s: %v", instance.Name(), instance.ID(), err))
+		return ctrl.Result{}, nil
```

Comment:
Do you need the current status here? If yes, you should call `GetInstanceByName` (or `GetInstanceStatusByName`) again.

Reply:
Also `instance` should be called `instanceStatus` :)

Reply:
I don't think we want to fetch again here. Were you thinking of some specific reason we might need to? We'd potentially race with the prior delete, so it might even be a bug to re-fetch.

Reply:
Why we are deleting the instance and after that checking for the instance status? Shouldn't we first check the status (i.e., to get the fips) and if that fails, we don't want to delete the instance. Otherwise we could have orphaned fips, right?

Reply:
We're creating a InstanceNetworkStatus here so we can get the FloatingIP. This doesn't go back to OpenStack; it's just doing some extra work on the previously returned server object. The only reason it can fail is json parsing. All we need to know is which FloatingIP to delete.

The previous code also had this failure mode, but it was in the return of DeleteInstance()->GetInstance()->serverToInstance()->GetIPFromInstance(). Moving that call up to the 'top level' in the controller is the primary purpose of this refactor, because it means we can pass additional arguments to NetworkStatus() so we can have more context. I don't think it currently matters for FloatingIP(), for for IP() it means we can pass OpenStackCluster as an argument here and use the additional context to determine the correct 'primary' IP from those returned by OpenStack.

Reply:
Incidentally, for robustness we should probably reverse the order of these deletions. The reason is that if deletion of the server succeeds but the deletion of the floating ip fails, the next reconcile will short-cut because there's no server and we'll leak the floating ip.

This patch is just a refactor, though, so I don't want to fix that here.

Reply:
That's exactly the point I was worried about. We should first delete the floating IPs and the instance right after the successful deletion.
But okay for me to fix that after this refactor PR.

Further, we could maybe improve the error message for the failed json parsing. `error getting network status for OpenStack instance` is a bit misleading, tbh.

Reply:
And starting errors with `error` is a bit redundant. But we have to fix that in the complete repo anyway